src/Nethermind/Nethermind.TxPool/Filters/FeeTooLowFilter.cs

```diff
@@ -13,36 +13,46 @@ namespace Nethermind.TxPool.Filters
     /// <summary>
     /// Filters out transactions where gas fee properties were set too low or where the sender has not enough balance.
     /// </summary>
-    internal class FeeTooLowFilter : IIncomingTxFilter
+    internal sealed class FeeTooLowFilter : IIncomingTxFilter
     {
         private readonly IChainHeadSpecProvider _specProvider;
         private readonly IChainHeadInfoProvider _headInfo;
         private readonly TxDistinctSortedPool _txs;
         private readonly ILogger _logger;
 
-        public FeeTooLowFilter(IChainHeadInfoProvider headInfo, TxDistinctSortedPool txs, ILogManager logManager)
+        public FeeTooLowFilter(IChainHeadInfoProvider headInfo, TxDistinctSortedPool txs, ILogger logger)
         {
             _specProvider = headInfo.SpecProvider;
             _headInfo = headInfo;
             _txs = txs;
-            _logger = logManager.GetClassLogger();
+            _logger = logger;
         }
 
         public AcceptTxResult Accept(Transaction tx, TxFilteringState state, TxHandlingOptions handlingOptions)
         {
+            bool isLocal = (handlingOptions & TxHandlingOptions.PersistentBroadcast) != 0;
+            if (isLocal)
+            {
+                return AcceptTxResult.Accepted;
+            }
+
             IReleaseSpec spec = _specProvider.GetCurrentHeadSpec();
-            UInt256 balance = state.SenderAccount.Balance;
-            UInt256 affordableGasPrice = tx.CalculateAffordableGasPrice(spec.IsEip1559Enabled, _headInfo.CurrentBaseFee, balance);
-            bool isNotLocal = (handlingOptions & TxHandlingOptions.PersistentBroadcast) != TxHandlingOptions.PersistentBroadcast;
+            bool isEip1559Enabled = spec.IsEip1559Enabled;
+            UInt256 affordableGasPrice = tx.CalculateGasPrice(isEip1559Enabled, _headInfo.CurrentBaseFee);
+            // Don't accept zero fee txns even if pool is empty as will never run
+            if (isEip1559Enabled && !tx.IsFree() && affordableGasPrice.IsZero)
+            {
+                Metrics.PendingTransactionsTooLowFee++;
+                if (_logger.IsTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, too low payable gas price with options {handlingOptions} from {new StackTrace()}");
+                return AcceptTxResult.FeeTooLow.WithMessage("Affordable FeePerGas of 0 rejected.");
+            }
 
-            if (isNotLocal
-                && _txs.IsFull()
-                && _txs.TryGetLast(out Transaction? lastTx)
+            if (_txs.IsFull() && _txs.TryGetLast(out Transaction? lastTx)
```

Comment:
I think we could have a lockless version of TryGetLast, just when something is added to that Set get the last value and save it to some field.

Reply:
`_txs.IsFull()` also takes a lock

Reply:
Added PR for that: https://github.com/NethermindEth/nethermind/pull/5360 can you try it?

Reply:
Merged

Reply:
> Added PR for that: #5360 can you try it?

Makes `FeeTooLowFilter` cheaper than `MalformedTxFilter` so switched the order

Reply:
<img width="575" alt="image" src="https://user-images.githubusercontent.com/1142958/222701690-56e3f715-0c74-4194-b2d8-6a89de4aae2a.png">